lms/views/api/sync.py

```diff
@@ -0,0 +1,35 @@
+import hashlib
+
+from pyramid.view import view_config
+
+from lms.models import HGroup
+
+
+@view_config(
+    route_name="api.sync", request_method="POST", renderer="json", permission="sync_api"
+)
+def sync(request):
+    canvas_api_svc = request.find_service(name="canvas_api_client")
+    lti_h_svc = request.find_service(name="lti_h")
+
+    context_id = request.json["course"]["context_id"]
+    custom_canvas_course_id = request.json["course"]["custom_canvas_course_id"]
+    tool_consumer_instance_guid = request.json["lms"]["tool_consumer_instance_guid"]
+
+    sections = canvas_api_svc.course_sections(custom_canvas_course_id)
+
+    groups = []
+
+    for section in sections:
+        # Generate a unique authority_provided_id for the section group.
+        hash_object = hashlib.sha1()
+        hash_object.update(tool_consumer_instance_guid.encode())
+        hash_object.update(context_id.encode())
+        hash_object.update(str(section["id"]).encode())
+        authority_provided_id = hash_object.hexdigest()
+
+        groups.append(HGroup(HGroup.group_name(section["name"]), authority_provided_id))
+
+    lti_h_svc.sync(groups)
+
+    return groups
```

Comment:
:man_shrugging: Aside from the 5 lines for generating `authority_provided_id` this view is just plugging together a bunch of already existing stuff:

* The sync API params that the frontend posted, as placed in the JavaScript config by `JSConfig`

* The Canvas API client's methods for getting sections from Canvas. There are actually three different methods for getting the sections in different ways under different circumstances, and this view is going to have to grow some logic for choosing between them. For now it's just using the all-of-the-courses-sections method all the time

* The `LTIHService`'s `sync()` method

* It then returns the groups to the frontend, which will need to pass them to the client

So far no logic in the view, just one loop. It will grow a little logic for selecting the right Canvas API client method to call, but I'm expecting it to remain simple enough for the code to get away with just all living in the view directly

Reply:
We should probably add a validation schema to the view for good measure, though it's not like the frontend is ever going to be calling it invalidly anyway